### PR TITLE
Compile contracts using solcjs or solc

### LINKED ISSUE #168: improve error message when solc is not installed
related https://github.com/paritytech/parity-bridge/issues/167

### LINKED ISSUE #169: Using solcjs instead of solc
`solc` and `solcjs` both can be used to compile smart contracts. One drawback of using solc is that solc is not easily available on many platforms. Specifically, I am talking about `centos`.

Yes it can be built from source. But I faced many difficulties while building solc. It required some libraries(boost-c-library), and some libraries also required some more libraries(z3). And these new libraries were also not easily available, so I had to build them(z3) too from source.

So, I am suggesting that we can use `solcjs`, which is more easily available on all platforms.

### PR DESCRIPTION
**Pull request by AyushyaChitransh**
This PR closes #168 and closes #169 and it also improves #170

 - First of all, try compiling with `solc`
 - If `solc` is not found, try using solcjs.
 - Panic when neither of compilation tools are available

### CODE REVIEW

**Comment by AyushyaChitransh** [comment]
@snd or @debris Can you please review this PR. It is ready from my side.

**Comment by snd** [comment]
i did not work last week. will look at it soon

File: bridge/build.rs [hunk 1]
```diff
@@ -40,14 +34,111 @@ fn main() {
...
+                        }
+                    }
+                }
             } else {
                 panic!("an error occurred when trying to spawn `solc`: {}", err);
             }
         }
     }
 }
+fn get_file_name(path: DirEntry) -> String {
```
Comment (snd) [inline bridge/build.rs]: i think it would be cleaner to inline this above

**Review by snd — COMMENTED** [review]
thanks for the contribution!

it would be great if you could try simplifying the deeply nested matches and if's. since most branches end with a panic! this could be written in an early-return (or in this case "early-panic") style which would make it more readable

**Comment by AyushyaChitransh** [comment]
@snd early panic cannot be possible. We are panicking only if solc and solcjs both are not available. Or if any process fails unexpectedly, example:

 - panic if !exit_status.success for solcjs. This means solcjs could not compile successfully on line [30](https://github.com/AyushyaChitransh/parity-bridge/blob/78ced125d5ed9a1d8db6ad9d1428c52dce44556d/bridge/build.rs#L30)

For any confusion regarding any panics, feel free to raise doubts. I'll be happy to update them with comments or try resolving those doubts.

File: bridge/build.rs [hunk 2]
```diff
@@ -40,11 +33,106 @@ fn main() {
...
+                        let prepend = "../compiled_contracts/";
+                        let paths = fs::read_dir("../compiled_contracts").unwrap();
+                        for path in paths {
+                            let _file_name: String = format!(
+                                "{}{}",
+                                prepend,
+                                path.unwrap().file_name().into_string().expect(
+                                    "error: the first argument is not a file\
+                                     system path representable in UTF-8.",
+                                )
```
Comment (debris) [inline bridge/build.rs]: please use `Path` instead. e.g. `Path::new(prepend).join(path.unwrap().file_name())` or maybe even `path.unwrap().path()`

File: bridge/build.rs [hunk 3]
```diff
@@ -40,11 +33,106 @@ fn main() {
...
+                        let solc_version = output_string.lines().last().unwrap();
+                        println!("cargo:rustc-env=SOLC_VERSION={}", solc_version);
+                        // overwrite option is not available in solcjs
+                        // so we better remove old compiled contracts
+                        match fs::remove_dir_all("../compiled_contracts") {
+                            Ok(()) => {
+                                println!("Removed old files");
+                            }
+                            Err(err) => {
+                                println!("Files not removed: {}", err);
```
Comment (debris) [inline bridge/build.rs]: using `println!` for debugging is usually a bad pattern. Especially in `build.rs`

**Review by debris — COMMENTED** [review]
This function is still very complex. Can you simplify it further by separating functions `compile_using_solc` and `compile_using_solcjs`. Also removing those nested matches and if/else statements would help a lot with code readibility. It's difficult to find the and of the block now :)

**Comment by AyushyaChitransh** [comment]
Tests are failing now due to timeout. Is this because we are using party version 1.10? Because previously, similar tests were working on using parity version 1.9.7

**Comment by snd** [comment]
seems like tests are failing because the code is not formatted with rustfmt. could you run `cargo fmt`, add, commit and try again? we've setup CI to enforce the codebase is always formatted with `cargo fmt`

**Comment by snd** [comment]
i think it makes sense to move this into a separate crate since the same functionality is also present in https://github.com/paritytech/sol-rs.

**Comment by snd** [comment]
i'm going to create a small crate that makes it easy for any project to interact with / compile sol files with solc and/or solcjs

**Comment by snd** [comment]
started working on said crate over here https://github.com/snd/rust_solc/.
it will use the JSON I/O format that's common to solc/solcjs.
this will enable compiling solidity from rust using solc/solcjs without touching the filesystem.
i think this is overall a cleaner and more reusable solution.
it also opens up things like https://github.com/paritytech/sol-rs/issues/13 without touching the filesystem

**Comment by snd** [comment]
wasn't able to get the JSON I/O to work just yet because it's broken on solcjs https://github.com/ethereum/solc-js/issues/126.

i got https://github.com/snd/rust_solc to work with file I/O though.

we should now be able to replace the solc compilation code in this PR by a simple

```rust
solc::compile_dir("../contracts", "../compiled_contracts").unwrap();
```

@AyushyaChitransh do you want to modify this PR accordingly or open another PR? let me know. otherwise i'll do it in another PR

**Comment by AyushyaChitransh** [comment]
Let me rephrase what I understood:

You are suggesting we use rust_solc. And on using that, we would not require to use `solc` or `solcjs`. 

If this is the case, lets have another PR from your side and close this one.

**Comment by snd** [comment]
> You are suggesting we use rust_solc. And on using that, we would not require to use solc or solcjs.

rust_solc is just a crate that shells out to solc or solcjs

> If this is the case, lets have another PR from you and close this one.

sounds good

**Comment by AyushyaChitransh** [comment]
Am still unable to compile. Facing solc errors. 

`thread 'main' panicked at '`solc --version` failed to run. run it yourself to verify. file an issue if this persists: Os { code: 2, kind: NotFound, message: "No such file or directory" }', bridge/build.rs:36:69`
